clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -163,6 +172,68 @@ public void testPartitionerClose() throws Exception {
         }
     }
 
+    @Test
+    public void shouldCloseProperlyAndThrowIfInterrupted() throws Exception {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(ProducerConfig.PARTITIONER_CLASS_CONFIG, MockPartitioner.class.getName());
+        props.setProperty(ProducerConfig.BATCH_SIZE_CONFIG, "1");
+
+        Time time = new MockTime();
+        Cluster cluster = TestUtils.singletonCluster("topic", 1);
+        Node node = cluster.nodes().get(0);
+
+        Metadata metadata = new Metadata(0, Long.MAX_VALUE, true);
+        metadata.update(cluster, Collections.<String>emptySet(), time.milliseconds());
+
+        MockClient client = new MockClient(time, metadata);
+        client.setNode(node);
+
+        final Producer<String, String> producer = new KafkaProducer<>(
+            new ProducerConfig(ProducerConfig.addSerializerToConfig(props, new StringSerializer(), new StringSerializer())),
+            new StringSerializer(), new StringSerializer(), metadata, client);
+
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        final AtomicReference<Exception> closeException = new AtomicReference<Exception>();
+        try {
+            Future<?> future = executor.submit(new Runnable() {
+                @Override
+                public void run() {
+                    producer.send(new ProducerRecord<>("topic", "key", "value"));
+                    try {
+                        producer.close();
+                        fail("Close should block and throw.");
+                    } catch (Exception e) {
+                        closeException.set(e);
+                    }
+                }
+            });
+
+            // Close producer should not complete until send succeeds
+            try {
+                future.get(100, TimeUnit.MILLISECONDS);
+                fail("Close completed without waiting for send");
+            } catch (java.util.concurrent.TimeoutException expected) { /* ignore */ }
+
+            // Ensure send has started
+            client.waitForRequests(1, 1000);
+
+            assertTrue("Close terminated prematurely", future.cancel(true));
+
+            // Make sure that close task completes and another task can be run on the single threaded executor
+            executor.submit(new Runnable() {
```

Comment:
Couldn't we assert that the future is completed instead?

Reply:
Cannot follow. Do you mean L221 ?

Reply:
I mean it's odd to submit an empty job to the executor in order to verify progress. Why not call `get` on the close future itself.

Reply:
I see. I copied this pattern from `KafkaCosumerTest#shouldCloseProperlyAndThrowIfInterrupted`.  Was added here: https://github.com/apache/kafka/commit/f72203ee9223d3b724ee67bdad9912612dd72f63

Not sure what the reasoning was. Maybe something we overlook. \cc @rajinisivaram

Reply:
I dont remember why I did that. Looks like `future.get` would be better.

Reply:
Thanks @rajinisivaram ! Will update in both tests.

Reply:
Ok. Tries this. But can't replace it; it's correct and needed.

It's a synchronization barrier to avoid a race condition. After `future#cancel()`, `get()` immediately throws, and `isDone()` just returns `true`. Thus, `closeException` might not have been set yet when we do the assert as the `Runnable` was not scheduled yet. A `sleep(500)` also makes the test pass, but it's less reliable and might introduce some flakiness. Thus, we should keep this IMHO.

Will update the comment in both tests thought.

Reply:
Thanks for the explanation. Perhaps a more straightforward solution might be to use a latch which is set in a finally block inside the close runnable?